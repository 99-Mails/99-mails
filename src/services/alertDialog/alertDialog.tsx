import {
  Button,
  AlertDialog,
  AlertDialogBody,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogContent,
  AlertDialogOverlay,
} from "@chakra-ui/react";

export default function AlertDialogWrapper({
  isOpen,
  cancelRef,
  onClick,
  onClose,
  onCancel,
  header,
  body,
  isLoading,
  isCancelable,
}: {
  isOpen: boolean;
  cancelRef: any;
  onClick: Fn;
  onClose: Fn;
  onCancel: Fn;
  header: string;
  body: string;
  isLoading: boolean;
  isCancelable: boolean;
}) {
  return (
    <AlertDialog
      data-testid="alert-dialog-root"
      isCentered={true}
      isOpen={isOpen}
      leastDestructiveRef={cancelRef}
      onClose={onClose}
    >
      <AlertDialogOverlay>
        <AlertDialogContent>
          <AlertDialogHeader
            data-testid="alert-dialog-header"
            fontSize="lg"
            fontWeight="bold"
          >
            {header}
          </AlertDialogHeader>

          <AlertDialogBody data-testid="alert-dialog-body">
            {body}
          </AlertDialogBody>

          <AlertDialogFooter>
            <Button
              ref={cancelRef}
              isDisabled={isCancelable}
              data-testid="alert-dialog-deny"
              onClick={() => onCancel()}
            >
              Cancel
            </Button>
            <Button
              data-testid="alert-dialog-accept"
              colorScheme="red"
              onClick={() => onClick(true)}
              ml={3}
              isLoading={isLoading}
            >
              Delete
            </Button>
          </AlertDialogFooter>
        </AlertDialogContent>
      </AlertDialogOverlay>
    </AlertDialog>
  );
}
