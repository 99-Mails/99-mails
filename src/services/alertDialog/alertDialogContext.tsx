import { Portal, useDisclosure } from "@chakra-ui/react";
import { createContext, useEffect, useRef, useState } from "react";
import AlertDialogWrapper from "./alertDialog";

interface IDialogContext {
  isOpen: boolean;
  onOpen?: Fn;
  onCancel?: Fn;
  onClose?: Fn;
  getConfirmation?: Fn;
  setButtonLoading?: Fn;
  setHeader?: Fn;
  setBody?: Fn;
  signal?: AbortSignal;
}

export const DialogContext = createContext<IDialogContext>({
  isOpen: false,
});

// TODO: code smell
// TODO: improve typing
// TODO: should move component origin to a react-portal
const DialogProvider = ({ children }) => {
  const abortControllerRef = useRef(new AbortController());

  const { isOpen, onOpen, onClose } = useDisclosure();
  const [loading, setButtonLoading] = useState(false);
  const [header, setHeader] = useState("");
  const [body, setBody] = useState("");
  const [resolver, setResolver] = useState({ resolve: null });

  const cancelRef = useRef();

  const createPromise = () => {
    let resolver: Fn;
    return [
      new Promise((resolve) => {
        resolver = resolve;
      }),
      resolver,
    ];
  };

  const getConfirmation = (arg: any) => {
    setBody(arg.body);
    setHeader(arg.header);
    onOpen();
    const [promise, resolve] = createPromise();
    setResolver({ resolve });
    return promise;
  };

  const onClick = async (status) => {
    resolver.resolve(status);
  };

  // TODO: fix re-fetch after an a request abortion
  const onCancel = () => {
    abortControllerRef.current.abort();
    setButtonLoading(false);
    setBody("");
    setHeader("");
    onClose();
  };

  // TODO: not working!
  // useEffect(() => {
  //   if (abortControllerRef.current.signal.aborted) {
  //     onClose();
  //     setButtonLoading(false);
  //   }
  // }, [abortControllerRef.current.signal.aborted]);

  const value = {
    isOpen,
    onOpen,
    onCancel,
    onClose,
    getConfirmation,
    setButtonLoading,
    setHeader,
    setBody,
    signal: abortControllerRef.current.signal,
  };

  return (
    <DialogContext.Provider value={value}>
      {children}
      <Portal>
        <AlertDialogWrapper
          body={body}
          cancelRef={cancelRef}
          header={header}
          isOpen={isOpen}
          isLoading={loading}
          isCancelable={!loading}
          onClick={onClick}
          onCancel={onCancel}
          onClose={onClose}
        />
      </Portal>
    </DialogContext.Provider>
  );
};

export { DialogProvider };
