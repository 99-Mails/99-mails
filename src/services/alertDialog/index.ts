export * from "./alertDialog";
export * from "./alertDialogAdaptor";
