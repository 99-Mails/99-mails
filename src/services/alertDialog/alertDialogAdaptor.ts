import { useContext } from "react";
import { AlertDialogService } from "@/application/ports";
import { DialogContext } from "./alertDialogContext";

// TODO: codesmells
// TODO: fix typing - should enable strict mode asap
// TODO: implement async loading for affirmative button pass âœ…
export function useAlertDialog(): AlertDialogService {
  const alert = useContext(DialogContext);

  function closeDialog() {
    alert.setHeader("");
    alert.setBody("");
    alert.setButtonLoading(false);
    alert.onClose();
  }

  return {
    closeDialog,
    openDialog: (arg) => alert.getConfirmation(arg),
    cancelDialog: () => alert.onCancel(),
    isLoading: () => alert.setButtonLoading(true),
    stopLoading: () => alert.setButtonLoading(false),
    signal: alert.signal,
  };
}
