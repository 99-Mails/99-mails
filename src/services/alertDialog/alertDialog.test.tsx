import { it } from "vitest";
import { render } from "@/tests/test-utils";
import { DialogProvider } from "./alertDialog";
import { useAlertDialog } from "./alertDialogAdaptor";
import { renderHook, act } from "@testing-library/react";

// TODO: should write a test!
it.skip("should set the header correctly", async () => {
  // const { container } = render(<div />, {
  //   wrapper: ({ children }) => {
  //     return <DialogProvider>{children}</DialogProvider>;
  //   },
  // });

  const wrapper = ({ children }) => {
    return <DialogProvider>{children}</DialogProvider>;
  };

  // const dialog = useAlertDialog();

  // const action = await dialog.openDialog({
  //   body: "Are you sure?",
  //   header: "Delete Address",
  // });

  const { result } = renderHook(() => useAlertDialog(), { wrapper });

  act(() => {
    result.current.openDialog({
      body: "Are you sure?",
      header: "Delete Address",
    });
  });

  expect(result.current).toContain("Delete Address");
});
