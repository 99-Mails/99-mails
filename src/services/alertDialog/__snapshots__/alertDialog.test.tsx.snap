// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `[Function]`;
